torchtune/modules/transformer.py

```diff
@@ -249,15 +424,27 @@ def forward(
             # in most cases input_pos_len should be 1
             mask = self.causal_mask[None, input_pos]
 
-        for layer in self.layers:
+        hidden = []
+        for i, layer in enumerate(self.layers):
+            if i in self.output_hidden_states:
+                hidden.append(h)
             # shape: [b, s, d]
-            h = layer(h, mask=mask, input_pos=input_pos)
+            h = layer(
+                h,
+                mask=mask,
+                encoder_input=encoder_input,
+                encoder_mask=encoder_mask,
+                input_pos=input_pos,
+            )
 
         # shape: [b, s, d]
         h = self.norm(h)
 
         # shape: [b, s, out_dim] - out_dim is usually the vocab size
         output = self.output(h).float()
+
+        # Output list if hidden states are requested, otherwise just the output
+        output = output if not hidden else [*hidden, output]
```

Comment:
Generally, I don't think I like the output type being dynamically determined. I'd much rather prefer if we always returned a list and and had output[-1] refer to the output. Otherwise the downstream users will always need to add checks related to the type. Ideally this isn't a list but a dataclass

Reply:
+ 1

Reply:
After some external discussion, I'm going to leave this like this for now so that the default behavior of the model doesn't change through the library. And I'll revisit this in a separate PR.

Reply:
I'll trade you an "Approve" for a "TODO" in the docstring which leaves a reminder to clean this up :)